### ISSUE TITLE
KOReader does a full refresh whenever anything changes on the screen (including time) when image is visible

### ISSUE THREAD

**Issue by lightmaster** [koreader/koreader#12218]
* KOReader version: 2024.07-25
* Device: Kobo Libra Colour

#### Issue

Any time an epub is open with an image on screen, the screen will do a full refresh at least once, even if a menu is blocking the image from view. The setting `Always flash on pages with images` will decrease the time it takes for the extra refresh to occur, but it will still do it regardless. I also noticed that even when all of the screen flash settings are disabled (such that it shouldn't flash the screen automatically at all), it will still flash every time you flip to a page with an image.

It does not appear to flash an extra time when viewing a CBZ file.

Ok, I think I've figured out the trigger. Every time any changes on the screen, even the time or the "Frontlight enabled/disabled" popup, when an image is on the screen in an epub, the entire screen will do a full refresh. It doesn't do it in a CBZ file since the time on mine is in the Alt Status Bar which isn't visible in a CBZ file. (Side note, would be great to get the option to enable the Alt Status Bar in a CBZ file) It does start doing it in a CBZ file if I have the menu open so the time is visible, but the time in the menu doesn't change automatically, only when you switch to a different menu, so it's really odd that it's flashing the whole screen when the menu's time is visible.

This is the first time I've read a graphic novel or comic on my Kobo, so I don't know if it's always done this or if it was something that changed recently. Either way, I don't think it should be doing a full refresh when it doesn't need to, and instead should only flash the little bit of the screen that did change. Time probably doesn't need a refresh at all, or at least it should be controlled by the option "Flash buttons and menu items".

#### Steps to reproduce

1. Open epub with images
2. Turn on Time in the status bar
3. Flip to page that has an image on it.

##### `crash.log` (if applicable)
[crash.log](https://github.com/user-attachments/files/16370920/crash.log)

**Comment by poire-z** [comment]
There's an option (in both Alt status bar and for the footer under Configure items) for "Auto refresh".
Can you see if disabling them both, or one and not the other, solves some issue? (It's possible a region update at the top + one at the bottom get merged into one for the full screen).
Anyway, pinging our auto-refreshed @zwim :)

**Comment by lightmaster** [comment]
Disabling both auto refreshes does seem to keep it from refreshing frequently.

I'm not sure if it has anything to do with top and bottom region updates being merged together, because clicking the bottom left corner to turn the front light on/off will cause the entire screen to refresh, even though the front light popup is only showing up in a small area at the top of the screen.

**Comment by poire-z** [comment]
May be you have the footer item "Frontlight level" enabled ? So notification at top + value at bottom = merge to fullscreen?

**Comment by NiLuJe** [comment]
That's a quirk of how color refreshes are currently handled.

I'm waiting on the kernel sources to confirm a hunch that could allow us to do things in a slightly less heavy-handed way.

**Comment by NiLuJe** [comment]
("need more info" from upstream, for once, not the OP ;p)

**Comment by NiLuJe** [comment]
> clicking the bottom left corner to turn the front light on/off will cause the entire screen to refresh

That's because of the frontlight notification, in addition to the quirk I mentioned earlier.

Dismissing a popup means we need to repaint the visible widgets underneath it. In the Reader, that's ReaderUI, and it's full-screen.
Now, usually, the *region* of the repaints triggered by a dismiss are ignored.
This is not the case for dithered/color widgets, though, and as it's full-screen, the refresh is full-screen.

And because color currently require a flashing refresh, it's a flashing full-screen refresh ;).

c.f., https://github.com/koreader/koreader/blob/477d6865f14731d63ef4e14e07090bd5382277f9/frontend/ui/uimanager.lua#L1127-L1135

**Comment by NiLuJe** [comment]
And there another quirk related to the thing I'm hoping we might be able to get rid of, which means that we currently request color waveform modes via `setDirty(nil, "color")`, and that has a tendency to also imply a full-screen refresh ;).

**Comment by lightmaster** [comment]
> May be you have the footer item "Frontlight level" enabled ? So notification at top + value at bottom = merge to fullscreen?

I did, but it still did it after removing brightness level from the bottom status bar.

@NiLuJe So if I understand that correctly, when the popup goes away it has to refresh the whole screen but y'all are looking into a better method? Also, in the case of the time changing, does that really need a refresh of the whole screen? I've tried disabling all settings for Full Refresh Rate and the flash screen settings under E-ink Settings, and when the Alt Status bar is set to Auto Refresh and the time is show, it still triggeres a full screen refresh. On a page with no images, the time changing does not trigger a full screen refresh.

**Comment by lightmaster** [comment]
> And because color currently requires a flashing refresh, it's a flashing full-screen refresh ;).

idk if it matters, but I did try disabling "Color rendering" and it still did a full screen refresh when the time changed"

**Comment by NiLuJe** [comment]
> idk if it matters, but I did try disabling "Color rendering" and it still did a full screen refresh when the time changed"

That won't change anything, it's based on the device capabilities, not the current rendering mode.

(Which, eeeeh, not great, but if I have my way, we won't be able to tell anyway, so not worth doing it "right" now just to break it later).

**Comment by NiLuJe** [comment]
> we won't be able to tell anyway

Oh, it's in Screen, so we actually might.

**Comment by NiLuJe** [comment]
Oh, I know why AltStatusBar is weird: because it's weird.

It's not a dedicated widget, it's CRe. As such, it's actually part of the Reader page, so it's painted by ReaderView. Which means it goes through the `setDirty(nil, "color")` of `ReaderView:paintTo`, because it *is* technically part of the document page. -_-".

**Comment by NiLuJe** [comment]
Note that auto-refreshes of the status bar *should* be unaffected, though. Only dismissing it should trip a (spurious) full-screen refresh.

(Reminder: status: bottom; alt: top).

**Comment by lightmaster** [comment]
> Note that auto-refreshes of the status bar _should_ be unaffected, though.

Time in the status bar (bottom) and set the status bar to auto refresh, and it did not trigger a full refresh.

> because it _is_ technically part of the document page

ooh, that explains why the Alt Status Bar isn't an option when reading a CBZ file.

**Comment by lightmaster** [comment]
Just got a chance to try out the latest Nightly with the fix. Everything seems to be good now. With "Always flash on pages with images" turned off, I'm only seeing the screen flash on the actually images, not the text around the images. Time auto updating in the Alt Status bar isn't causing any flash, and the Frontlight popup isn't causing any flashing.
